app/src/main/java/com/epfl/drawyourpath/database/FireDatabase.kt

```diff
@@ -38,76 +71,199 @@ class FireDatabase : Database() {
         return future
     }
 
-    override fun updateUsername(username: String, userId: String): CompletableFuture<Boolean> {
-        TODO("This will be implmented during next task when I will clean the database organisation")
+    override fun updateUsername(username: String): CompletableFuture<Boolean> {
+        val future = CompletableFuture<Boolean>()
+
+        val userId = getUserId()
+        if(userId == null){
+            future.completeExceptionally(java.lang.Error("The userId can't be null !"))
+        }else {
+            //obtain the past username form the userId
+            getUsernameFromUserId(userId).thenAccept { pastUsername ->
+                if (pastUsername == null) {
+                    future.completeExceptionally(java.lang.Error("Impossible to find the past username !"))
+                } else {
+                    //update the link username to userId and the username on the userAccount
+                    setUsername(username).thenAccept { isSetUsername ->
+                        if (isSetUsername) {
+                            //remove the past username from the link username/userId
+                            accessUsernameToUserIdFile().child(pastUsername).removeValue()
+                                .addOnSuccessListener { future.complete(true) }
+                                .addOnFailureListener {
+                                    future.completeExceptionally(
+                                        java.lang.Error(
+                                            "Impossible to remove the past username link !"
+                                        )
+                                    )
+                                }
+                        } else {
+                            future.completeExceptionally(java.lang.Error("Impossible to set this username !"))
+                        }
+                    }
+                }
+            }
+        }
+        return future
+    }
+
+    override fun setUsername(username: String): CompletableFuture<Boolean> {
+        val future = CompletableFuture<Boolean>()
+        //TODO:Add security rules to database
+        val userId = getUserId()
+        if(userId == null){
+            future.completeExceptionally(java.lang.Error("The userId can't be null !"))
+        }else{
+            //check if the username is available
+            isUsernameAvailable(username).thenAccept { isAvailable ->
+                if (isAvailable) {
+                    //add the link between the username and the userId
+                    val usernameToUserId = HashMap<String, String>()
```

Comment:
Doing two transactions here could result in partial state being saved to the database in case of connection failure. Realtime Database allows to do that in a single operation by using a complete path of the data.

Reply:
I understand but we want to make sur that the previous transaction its made bedaiore begin a new one. And in any case an error will be throw in this case

Reply:
But then shouldn't we rollback the update? Because the name will be updated in the profile but not in the mapping.

Reply:
After further discussion, we decided to go that way for now and implement the other method at a later time.